numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -0,0 +1,3035 @@
+            buf = (char *)os.buf;
+        }
+
+        memcpy(buf, s1.buf, s1.size);
+        memcpy(buf + s1.size, s2.buf, s2.size);
+
+        // clean up temporary in-place buffer
+        if (s1descr == odescr || s2descr == odescr) {
+            if (NpyString_pack(oallocator, ops, buf, newsize) < 0) {
+                npy_gil_error(PyExc_MemoryError,
+                          "Failed to pack output string in add");
+                goto fail;
+            }
+
+            PyMem_RawFree(buf);
+        }
+
+    next_step:
+        in1 += in1_stride;
+        in2 += in2_stride;
+        out += out_stride;
+    }
+    NpyString_release_allocators(3, allocators);
+    return 0;
+
+fail:
+    NpyString_release_allocators(3, allocators);
+    return -1;
+}
+
+static int
+maximum_strided_loop(PyArrayMethod_Context *context, char *const data[],
+                     npy_intp const dimensions[], npy_intp const strides[],
+                     NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_StringDTypeObject *in1_descr =
+            ((PyArray_StringDTypeObject *)context->descriptors[0]);
+    PyArray_StringDTypeObject *in2_descr =
+            ((PyArray_StringDTypeObject *)context->descriptors[1]);
+    npy_intp N = dimensions[0];
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *out = data[2];
+    npy_intp in1_stride = strides[0];
+    npy_intp in2_stride = strides[1];
+    npy_intp out_stride = strides[2];
+
+    npy_string_allocator *allocators[3] = {};
+    NpyString_acquire_allocators(3, context->descriptors, allocators);
+    npy_string_allocator *in1_allocator = allocators[0];
+    npy_string_allocator *in2_allocator = allocators[1];
+    npy_string_allocator *out_allocator = allocators[2];
+
+    while (N--) {
+        const npy_packed_static_string *sin1 = (npy_packed_static_string *)in1;
+        const npy_packed_static_string *sin2 = (npy_packed_static_string *)in2;
+        npy_packed_static_string *sout = (npy_packed_static_string *)out;
+        if (_compare(in1, in2, in1_descr, in2_descr) > 0) {
+            // if in and out are the same address, do nothing to avoid a
+            // use-after-free
+            if (in1 != out) {
+                if (free_and_copy(in1_allocator, out_allocator, sin1, sout,
+                                  "maximum") == -1) {
+                    goto fail;
+                }
+            }
+        }
+        else {
+            if (in2 != out) {
+                if (free_and_copy(in2_allocator, out_allocator, sin2, sout,
+                                  "maximum") == -1) {
+                    goto fail;
+                }
+            }
+        }
+        in1 += in1_stride;
+        in2 += in2_stride;
+        out += out_stride;
+    }
+
+    NpyString_release_allocators(3, allocators);
+    return 0;
+
+fail:
+    NpyString_release_allocators(3, allocators);
+    return -1;
+}
+
+static int
+minimum_strided_loop(PyArrayMethod_Context *context, char *const data[],
```

Comment:
The `minimum` and `maximum` strided loops are also a bit of `auxdata` away of being able to share the same code... Named them in the task list.